### PR TITLE
dev: add integration test environment for `eth-rpc`

### PR DESCRIPTION
**Pull request by ftupas**
Time spent on this PR: 0.5

Resolves: #365 

# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing

# What is the new behavior?

- Add integration test environment for Kakarot RPC with the following
  - TestSequencer
  - Deployed Kakarot contracts
  - Kakarot RPC server
- Move `test_utils` so it can be used in other crates

# Does this introduce a breaking change?

- [ ] Yes
- [X] No

### CODE REVIEW

File: crates/eth-rpc/tests/utils.rs [hunk 1]
```diff
@@ -35,3 +47,77 @@ pub async fn setup_mock_eth_rpc() -> KakarotEthRpc<JsonRpcClient<MockJsonRpcTran
...
+/// # Returns
+///
+/// This function returns a Result containing a tuple with the server's address and a handle to
+/// stop the server upon successful execution.
+///
+/// The function may return an Err variant of eyre::Report if there are issues with deploying the
+/// Kakarot contracts, creating the clients, or running the RPC server.
+///
+/// # Example
+/// ```ignore
```
Comment (Eikix) [inline crates/eth-rpc/tests/utils.rs:75]: I think this ``` is not closed

**Review by Eikix — COMMENTED** [review]
Very cool PR! A minor comment but lgtm, letting @ClementWalter or @jobez take a look

File: crates/eth-rpc/tests/utils.rs [hunk 2]
```diff
@@ -1,11 +1,23 @@
...
 use kakarot_rpc_core::mock::mock_starknet::{all_fixtures, init_mock_client};
-use starknet::providers::JsonRpcClient;
+use kakarot_rpc_core::test_utils::constants::EOA_WALLET;
+use kakarot_rpc_core::test_utils::deploy_helpers::deploy_kakarot_system;
+use starknet::core::types::FieldElement;
+use starknet::providers::jsonrpc::HttpTransport as StarknetHttpTransport;
+use starknet::providers::{JsonRpcClient, JsonRpcClient as StarknetJsonRpcClient};
 
-/// Run wiremock to fake starknet rpc and then run our own `kakarot_rpc_server`.
+/// Run wiremock to fake Starknet rpc and then run our own `kakarot_rpc_server`.
```
Comment (ClementWalter) [inline crates/eth-rpc/tests/utils.rs:20]: I thought that we removed `wiremock`, so probably needs an update

File: crates/eth-rpc/tests/utils.rs [hunk 3]
```diff
@@ -35,3 +47,78 @@ pub async fn setup_mock_eth_rpc() -> KakarotEthRpc<JsonRpcClient<MockJsonRpcTran
...
+///    // Create a TestSequencer.
+///    let test_sequencer = Arc::new(TestSequencer::new());
+///
+///    // Set up the Kakarot RPC integration environment.
+///    let (server_addr, server_handle) = setup_kakarot_rpc_integration_env(&test_sequencer).await.unwrap();
+///    
+///    // Query whatever eth_rpc endpoints
+///     
+///    // Dont forget to close server at the end.
+///    server_handle.stop().expect("Failed to stop the server");
```
Comment (ClementWalter) [inline crates/eth-rpc/tests/utils.rs:92]: can't we have this automatically done in `Drop` for example?

File: crates/eth-rpc/tests/utils.rs [hunk 4]
```diff
@@ -35,3 +47,78 @@ pub async fn setup_mock_eth_rpc() -> KakarotEthRpc<JsonRpcClient<MockJsonRpcTran
...
+///
+/// }
+/// ```
+pub async fn setup_kakarot_rpc_integration_env(
+    starknet_test_sequencer: &Arc<TestSequencer>,
+) -> Result<(SocketAddr, ServerHandle), eyre::Report> {
+    // Define expected funded amount.
```
Comment (ClementWalter) [inline crates/eth-rpc/tests/utils.rs:99]: why is this "expected"?

File: crates/eth-rpc/tests/rpc.rs [hunk 5]
```diff
@@ -1,5 +1,5 @@
 #![recursion_limit = "1024"]
-mod utils;
+pub mod utils;
```
Comment (ClementWalter) [inline crates/eth-rpc/tests/rpc.rs:2]: why do we need to `pub` this now?

File: crates/eth-rpc/tests/integration.rs [hunk 6]
```diff
@@ -0,0 +1,40 @@
+pub mod utils;
```
Comment (ClementWalter) [inline crates/eth-rpc/tests/integration.rs:1]: why `pub` ?

File: crates/eth-rpc/tests/integration.rs [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    #[tokio::test]
+    async fn test_get_block_rpc() {
+        // Setup the test sequencer
+        let starknet_test_sequencer = Arc::new(construct_kakarot_test_sequencer().await);
+
+        // Deploy the Kakarot contracts and start the Kakarot RPC server
+        let (server_addr, server_handle) = setup_kakarot_rpc_integration_env(&starknet_test_sequencer)
+            .await
+            .map_err(|e| {
+                println!("Error setting up Kakarot RPC server: {}", e);
```
Comment (ClementWalter) [inline crates/eth-rpc/tests/integration.rs:23]: not sure to understand why we don this `map_err`; is this to sort of do a manual logging? ain't there any cleaner way to do so?

File: crates/eth-rpc/tests/integration.rs [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+                e
+            })
+            .unwrap();
+
+        // Try to run the test
```
Comment (ClementWalter) [inline crates/eth-rpc/tests/integration.rs:28]: Do we "try to run" the test, or do we actually run it? ;)

File: crates/eth-rpc/tests/integration.rs [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        let provider = EthersHttp::from_str(format!("http://localhost:{}", server_addr.port()).as_ref()).unwrap();
+        let block_number: U64 =
+            ethers::prelude::JsonRpcClient::request(&provider, "eth_blockNumber", ()).await.unwrap();
+        let params = (block_number, true);
+        let block: std::result::Result<EthersBlock<EthersH256>, _> =
+            ethers::prelude::JsonRpcClient::request(&provider, "eth_getBlockByNumber", params).await;
```
Comment (ClementWalter) [inline crates/eth-rpc/tests/integration.rs:34]: not sure about what we want to test here? why do we call these two routes especially? I guess that this is in preparation for hive tests right? but this test confuses we

File: crates/eth-rpc/tests/integration.rs (hunk 8 above)
Comment (ftupas) [reply crates/eth-rpc/tests/integration.rs:28]: Indeed we actually run it! haha 😅

File: crates/eth-rpc/tests/integration.rs (hunk 9 above)
Comment (ftupas) [reply crates/eth-rpc/tests/integration.rs:34]: I wanted to test that by calling `eth_getBlockByNumber` endpoint, we are getting a block given all systems being deployed, so for that I'm passing the `block_number` as parameter which I get from the `eth_blockNumber` endpoint

File: crates/eth-rpc/tests/utils.rs (hunk 2 above)
Comment (ClementWalter) [reply crates/eth-rpc/tests/utils.rs:20]: mock yes, `wiremock` no

File: crates/eth-rpc/tests/integration.rs (hunk 6 above)
Comment (ClementWalter) [reply crates/eth-rpc/tests/integration.rs:1]: but in the `rpc.rs` case for example, it was not `pub` before and it was working, and you added `pub` as well. I still don't see why we would need to add this. We may need to, but I'd like to understand first

**Review by ClementWalter — COMMENTED** [review]
we need to understand this `pub` issue

File: crates/eth-rpc/src/test_utils.rs [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+///    let (server_addr, server_handle) = setup_kakarot_rpc_integration_env(&test_sequencer).await.unwrap();
+///    
+///    // Query whatever eth_rpc endpoints
+///     
+///    // Dont forget to close server at the end.
+///    server_handle.stop().expect("Failed to stop the server");
+///
+/// }
+/// ```
+pub async fn setup_kakarot_rpc_integration_env(
```
Comment (ftupas) [inline crates/eth-rpc/src/test_utils.rs:63]: Finally moved this to `test_utils` so it can be used in integration tests i.e. conformance test

Comment (ftupas) [reply crates/eth-rpc/src/test_utils.rs:63]: No more `pub` issue on the `utils.rs` file

**Review by ClementWalter — APPROVED** [review]

**Pull request merged**